os/dev/spi.c

```diff
@@ -0,0 +1,238 @@
+/*
+ * Copyright (c) 2016-2017, Yanzi Networks.
+ * Copyright (c) 2017, University of Bristol - http://www.bristol.ac.uk/
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ * 3. Neither the name of the copyright holder nor the names of its
+ *    contributors may be used to endorse or promote products derived
+ *    from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDER AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+ * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+/*---------------------------------------------------------------------------*/
+/**
+ * \addtogroup spi-hal
+ * @{
+ *
+ * \file
+ *     Implementation of the platform-independent aspects of the SPI HAL
+ */
+/*---------------------------------------------------------------------------*/
+#include <dev/spi.h>
+#include "contiki.h"
+#include <stdint.h>
+#include <stdbool.h>
+/*---------------------------------------------------------------------------*/
+spi_status_t
+spi_acquire(spi_device_t *dev)
+{
+  spi_status_t r;
+
+  if(dev == NULL || dev->spi_controller >= SPI_CONTROLLER_COUNT) {
+    return SPI_DEV_STATUS_EINVAL;
+  }
+
+  /* lock the bus */
+  r = spi_arch_lock(dev);
+  if(r != SPI_DEV_STATUS_OK) {
+    return r;
+  }
+
+  /* open the bus */
+  return spi_arch_open(dev);
+}
+/*---------------------------------------------------------------------------*/
+spi_status_t
+spi_release(spi_device_t *dev)
+{
+  spi_status_t r;
+
+  if(dev == NULL || dev->spi_controller >= SPI_CONTROLLER_COUNT) {
+    return SPI_DEV_STATUS_EINVAL;
+  }
+
+  /* unlock the bus */
+  r = spi_arch_close(dev);
+  if(r != SPI_DEV_STATUS_OK) {
```

Comment:
Hmm.. according to the documentation in `spi.h` line 314, `spi_arch_close()` fails only if the device that calls the function is the one which owns the lock. Please, do correct me if I am wrong, but if `spi_arch_close()` is implemented as specified, the current implementation would work identically to what you propose: if `spi_arch_close()` fails, the spi bus should remain locked, because someone else is currently using it. However, the scenario you describe would indeed appear if `spi_arch_close()` fails for other reasons.

I don't see any problem with merging them. Yet, I'd suggest to use function names that make this explicit - not to create ambiguity in future implementations. Perhaps, `spi_arch_lock_and_open()` and `spi_arch_close_and_unlock()`?